Replace large array in shr_expr_parser_mod with linked list

fvitt opened the pull request:
Passes:
SMS_D.f19_g16.X.hobart_intel
SMS_D_Ld1.f45_f45.FC5L45BGC.hobart_nag.cam-fire_emis 
SMS_D_Ld1.f19_f19.FC5L45BGC.hobart_intel.cam-fire_emis
SMS_D_Ld3.f45_g37_rx1.A.goldbach_nag 
as well as drv prealpha intel tests on yellowstone

fvitt committed:
reduce memory required to parse the MEGAN and fire emis namelists

Replace the large static array returned by shr_expr_parser_mod::shr_exp_parse
with a linked list.

  modified:   driver_cpl/shr/shr_expr_parser_mod.F90
  modified:   driver_cpl/shr/shr_fire_emis_mod.F90
  modified:   driver_cpl/shr/shr_megan_mod.F90

fvitt committed:
nullify pointers to appease nag compiler

 modified:   driver_cpl/shr/shr_expr_parser_mod.F90

jedwards4b commented:
Hi Francis,

I know that this list is intended to exist until the model run completes and so technically it doesn't need to be deallocated.   But I think that for completeness shr_expr_parser_mod.F90 should include a destroy method.

fvitt committed:
added a method to destroy linked list

The linked list returned by shr_exp_parse needs to exist
temporarily in shr_fire_emis_init and shr_megan_init so
the memory should be freed.  A new shr_exp_list_destory
method has been added for this purpose.

    modified:   driver_cpl/shr/shr_expr_parser_mod.F90
    modified:   driver_cpl/shr/shr_fire_emis_mod.F90
    modified:   driver_cpl/shr/shr_megan_mod.F90

fvitt commented:
Jim,

Thanks for taking a look at this.  I have added a destroy method.  The linked list needs to exist only during initialization of megan and fire_emis and should be destroyed to free up the memory.
